### PR TITLE
Concurrent writes; exclude_me; Goodbye knows session id

### PR DESCRIPTION
**Pull request by gngeorgiev**
- Added mutex to realm to prevent concurrent map writes
- Added exclude_me option to broker: https://tools.ietf.org/html/draft-oberstet-hybi-tavendo-wamp-02#section-13.5
- Goodbye now knows session id

### CODE REVIEW

File: message.go [hunk 1]
```diff
@@ -224,6 +224,7 @@ func (msg *Authenticate) MessageType() MessageType {
 // [GOODBYE, Details|dict, Reason|uri]
 type Goodbye struct {
 	Details map[string]interface{}
+	Request ID
```
Comment (beatgammit) [inline message.go:227]: This isn't up to [the spec](https://tools.ietf.org/html/draft-oberstet-hybi-tavendo-wamp-02#section-6.4).

It looks like you just wanted it for error reporting in client.go?

File: realm.go [hunk 2]
```diff
@@ -83,9 +86,13 @@ func (l *localClient) onLeave(session ID) {
 }
 
 func (r *Realm) handleSession(sess Session) {
+	r.Lock()
+	defer r.Unlock()
```
Comment (beatgammit) [inline realm.go:90]: This looks dangerous. I noticed you have different logic in your fork. I'll probably use that instead.

File: realm.go [hunk 3]
```diff
@@ -2,6 +2,8 @@ package turnpike
 
 import (
 	"fmt"
+	_ "log"
```
Comment (beatgammit) [inline realm.go:5]: What's the point of this?

**Comment by beatgammit** [comment]
Sorry for taking so long to get back to you on this, I've been a bit swamped at work. I'm happy to merge this if:
- remove the request ID in the Goodbye message to be up to spec (could do some magic in the reflection logic with a struct field tag)
- fix the issue I noted in the concurrent map writes
- clean up logic in exclude_me to be simpler (`(excludePublisher && sub != pub) || !excludePublisher` -> `sub != pub || !excludePublisher`)

Besides that, it looks good.

**Comment by mourad** [comment]
pulled this in as part of commit 62db20c

**Pull request closed without merging**